### PR TITLE
pypi: update setup.py and pyproject.toml for cibuildwheel

### PR DESCRIPTION
**Pull request by j2kun**
Corresponds to https://github.com/llvm/torch-mlir-release/pull/32

Cf. https://github.com/llvm/torch-mlir/discussions/4603

This can be tested locally with:

```
python3 -m venv venv
source venv/bin/activate
pip install cibuildwheel uv
CIBW_BUILD="cp312-manylinux_x86_64" cibuildwheel --platform linux
```

I also went ahead and added abi3 support for better cross-platform compatibility (in particular, the gLinux box I was testing it on could not install the wheel without abi3, though I didn't dig into why). However, that decision adds a constraint: LLVM pins the stable ABI to Python3.12, so that would limit this wheel to python3.12 or later. Is that OK? Python 3.11 is EOL on October 31, 2027, so there is an argument to be made that it would be useful to keep support for it around until then. But if LLVM as a whole agreed on python3.12 (I wouldn't know why) then maybe it's better to be consistent with that.

### CODE REVIEW

**Comment by j2kun** [comment]
I asked around the MLIR folks who work on the Python API, and [they mentioned](https://discord.com/channels/636084430946959380/642426447167881246/1524235894574284931) that the stable abi is required to be Python3.12 or newer to support nanobind. Since torch-mlir is actively using nanobind, I am inferring this forces torch-mlir to be Python3.12 or newer, so unless I hear otherwise, I won't change this PR to remove abi3.

I tried adding unstable wheels for Python 3.11, but I am not able to test it locally because the unstable wheel cannot be installed on my dev machine. (though the wheel built fine in cibuildwheel)

So the last open question (beyond normal review) is whether I should reinstate the unstable python 3.11 wheel.

**Comment by zjgarvey** [comment]
Taking a look today. The CI failure is fixed on main, so feel free to pull that in.

File: setup.py [hunk 1]
```diff
@@ -214,7 +223,7 @@ def run(self):
 
 class CMakeExtension(Extension):
     def __init__(self, name, sourcedir=""):
-        Extension.__init__(self, name, sources=[])
+        Extension.__init__(self, name, sources=[], py_limited_api=True)
         self.sourcedir = os.path.abspath(sourcedir)
```
Comment (zjgarvey) [inline setup.py]: The `_jit_ir_importer` path is not limited api, but is included by the unconditional tag here.

Maybe not super important because we shouldn't be releasing anything built with the jit-ir-importer anyway. It only needs to be built for the CI to auto-generate the ODS and abstract interp lib. I think discussion #4499 number 4. is essentially to remove this particular CI dependency and remove all the old code.

File: build_tools/smoke_test.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        self.linear = torch.nn.Linear(4, 4)
+        self.relu = torch.nn.ReLU()
+
+    def forward(self, x):
+        return self.relu(self.linear(x))
+
+
+module_to_compile = MLPModule()
+# Fix the weights
+torch.manual_seed(0)
```
Comment (zjgarvey) [inline build_tools/smoke_test.py]: nit: Why set a global manual seed here? We aren't doing numerics validation. Also, it's called after the module is initialized so the weights are already set.

**Review by zjgarvey — COMMENTED** [review]
I don't see any real benefit of hardcoding the stable api switches in `setup.py` since dependent scripts/jobs might still use older python versions. 

E.g., if we want to continue to release some older python version wheels for linux, having the option to turn off the stable api would be useful.

**Comment by j2kun** [comment]
> I don't see any real benefit of hardcoding the stable api switches in `setup.py` since dependent scripts/jobs might still use older python versions.
> 
> E.g., if we want to continue to release some older python version wheels for linux, having the option to turn off the stable api would be useful.

I'd like to try to avoid having a config option if it's not really needed. Would be it acceptable to use stable for Python 3.12 and later, and always use unstable for <=3.11? I.e., is there any reason someone would want to disable the stable abi for newer Python versions?

**Edit:** I tried this in ea3199e6, PTAL and I can add a proper switch if desired

File: build_tools/smoke_test.py (hunk 2 above)
Comment (j2kun) [reply build_tools/smoke_test.py]: Ah, thanks. I thought I was going to validate the result but then decided against it.

File: setup.py (hunk 1 above)
Comment (j2kun) [reply setup.py]: Acknowledged. IIUC there is nothing specific to do here (except as mentioned above, to make the stable abi depend on the version).

**Comment by j2kun** [comment]
After the most recent commit I was able to build all the linux wheels locally:

```
3 wheels produced in 29 minutes

  cp310-manylinux_x86_64: torch_mlir-0.0.1-cp310-cp310-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl 81.4 MB in 21 minutes, SHA256=7da537ed71595c3fa73731e8fd2962d2d3ea9abf7951837692bc7191c1f6e5b3
  cp311-manylinux_x86_64: torch_mlir-0.0.1-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl 81.4 MB in 2 minutes, SHA256=46c219a4a0e84ca7e4b75c0b5bb58e51f3c00accfa6c707a99a337e2c9e61060
  cp312-manylinux_x86_64: torch_mlir-0.0.1-cp312-abi3-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl 81.4 MB in 3 minutes, SHA256=ad767157171ef5e634c5c7c81ecf9865a2dc3dc8327140f4eca6692f25cc9337
```

File: pyproject.toml [hunk 3]
```diff
@@ -1,7 +1,60 @@
 [build-system]
-requires = ["setuptools", "wheel"]
+# These must be enough to drive setup.py's CMake build of the MLIR Python
+# bindings under build isolation (cibuildwheel). Kept in sync with the
+# [dependency-groups].build table below and
+# externals/llvm-project/mlir/python/requirements.txt (re-sync on LLVM bumps).
```
Comment (sahas3) [inline pyproject.toml:5]: Do you know what the failure mode will be if these become out-of-sync? Will it be easily deductible that the root cause is that these requirements are out-of-sync? 

Maybe as a follow-up a lint CI can be added that parses this file and llvm's requirements.txt to ensure the versions are in sync?

File: pyproject.toml [hunk 4]
```diff
@@ -1,7 +1,60 @@
...
+# object cache across the per-Python-version builds; CMake reads the
+# *_COMPILER_LAUNCHER cache vars from these env vars. CCACHE_DIR lives inside
+# the mounted /project so the host's actions/cache can persist it.
+[tool.cibuildwheel.linux]
+archs = "auto64"
+manylinux-x86_64-image = "manylinux_2_28"
+manylinux-aarch64-image = "manylinux_2_28"
+before-all = "dnf install -y --quiet epel-release || true && dnf install -y --quiet ccache || dnf install -y ccache"
+environment-pass = ["TORCH_MLIR_PYTHON_PACKAGE_VERSION"]
+environment = { TORCH_MLIR_ENABLE_ONLY_MLIR_PYTHON_BINDINGS = "1", TORCH_MLIR_ENABLE_JIT_IR_IMPORTER = "0", TORCH_MLIR_ENABLE_LTC = "0", CMAKE_GENERATOR = "Ninja", CMAKE_C_COMPILER_LAUNCHER = "ccache", CMAKE_CXX_COMPILER_LAUNCHER = "ccache", CCACHE_DIR = "/project/.ccache", CCACHE_MAXSIZE = "2G" }
```
Comment (sahas3) [inline pyproject.toml:50]: Surprised to see 2G cache size to be enough for the LLVM+torch-mlir build :)

File: build_tools/smoke_test.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        module = torch_mlir.fx.export_and_import(
+            module_to_compile, torch.ones(2, 4), output_type=backend
+        )
+        print(f"Compilation to {backend} successful!")
+        asm = module.operation.get_asm(large_elements_limit=10)
+        lines = asm.splitlines()
+        print(f"Generated ASM (first 10 lines of {backend}):")
+        for line in lines[:10]:
+            print("  ", line)
+        print("-" * 40)
```
Comment (sahas3) [inline build_tools/smoke_test.py]: nit: there doesn't seem to be any benefit for printing the output, should this be removed?

**Review by sahas3 — APPROVED** [review]
Thanks for this change. LGTM but please wait for @zjgarvey to approve as well before merging.

Also, apologies for the delay in reviewing.

File: pyproject.toml (hunk 3 above)
Comment (j2kun) [reply pyproject.toml:5]: The failure mode is that the cmake build will fail when running cibuildwheel, which will halt the release workflow.

This sounds like a good job for a pre-commit hook, I can add that in a followup.

File: pyproject.toml (hunk 4 above)
Comment (j2kun) [reply pyproject.toml:50]: I have not experimented with increasing this, but the wheels all build on my local system in 25 minutes, so it seems like there is little benefit to increasing it.

File: build_tools/smoke_test.py (hunk 5 above)
Comment (j2kun) [reply build_tools/smoke_test.py]: Done in eae3f2d342b2a554750519e2a3dc8f9c12e443d5

File: pyproject.toml (hunk 3 above)
Comment (j2kun) [reply pyproject.toml:5]: Added in #4655 (rebased over this pR)

**Review by zjgarvey — APPROVED** [review]
I was out sick, sorry for taking a bit to approve. LGTM.
